src/main/java/io/odpf/firehose/sink/bigquery/proto/ProtoUpdateListener.java

```diff
@@ -0,0 +1,104 @@
+        super(config.getInputSchemaProtoClass());
+        this.config = config;
+        this.bqClient = bqClient;
+        this.messageRecordConverterWrapper = messageRecordConverterWrapper;
+    }
+
+    public void update(Map<String, DescriptorAndTypeName> newDescriptors) {
+        onProtoUpdate("", newDescriptors);
+    }
+
+    @Override
+    public void onProtoUpdate(String url, Map<String, DescriptorAndTypeName> newDescriptors) {
+        log.info("stencil cache was refreshed, validating if bigquery schema changed");
+        try {
+            ProtoField protoField = new ProtoField();
+            protoField = protoMappingParser.parseFields(protoField, config.getInputSchemaProtoClass(), StencilUtils.getAllProtobufDescriptors(newDescriptors), StencilUtils.getTypeNameToPackageNameMap(newDescriptors));
+            updateProtoParser(protoField);
+        } catch (BigQueryException | IOException e) {
+            String errMsg = "Error while updating bigquery table on callback:" + e.getMessage();
+            log.error(errMsg);
+            throw new RuntimeException(errMsg, e);
+        }
+    }
+
+    // First get latest protomapping, update bq schema, and if all goes fine
+    // then only update beast's proto mapping config
+    private void updateProtoParser(final ProtoField protoField) throws IOException {
+        String protoMappingString = protoMappingConverter.generateColumnMappings(protoField.getFields());
+        List<Field> bqSchemaFields = protoMappingConverter.generateBigquerySchema(protoField);
+        addMetadataFields(bqSchemaFields);
+        bqClient.upsertTable(bqSchemaFields);
+        config.setProperty("SINK_BIGQUERY_PROTO_COLUMN_MAPPING", protoMappingString);
+        setProtoParser(config.getProtoColumnMapping());
+    }
+
+    private void addMetadataFields(List<Field> bqSchemaFields) {
+        List<Field> bqMetadataFields = new ArrayList<>();
+        String namespaceName = config.getBqMetadataNamespace();
+        if (namespaceName.isEmpty()) {
+            bqMetadataFields.addAll(BQField.getMetadataFields());
+        } else {
+            bqMetadataFields.add(BQField.getNamespacedMetadataField(namespaceName));
+        }
+
+        List<String> duplicateFields = getDuplicateFields(bqSchemaFields, bqMetadataFields).stream().map(Field::getName).collect(Collectors.toList());
+        if (duplicateFields.size() > 0) {
+            throw new RuntimeException(String.format("Metadata field(s) is already present in the schema. "
+                                                     + "fields: %s", duplicateFields));
+        }
+        bqSchemaFields.addAll(bqMetadataFields);
+    }
+
+    private void setProtoParser(Properties columnMapping) {
+        messageRecordConverterWrapper.setMessageRecordConverter(
+                new MessageRecordConverter(new RowMapper(columnMapping, config.getFailOnUnknownFields()),
```

Comment:
i see, here is where the wrapper is being used, because RowMapper is the only thing that changes on proto schema update, maybe we to set the value of row mapper rather that replace the whole record converter
or 
maybe rename Wrapper as Cache instead, because it behave like Cache, and the name will be more intentional

Reply:
yeah, naming can be changed.